From b13b1263965f690c92c02091081dd017e527c8fe Mon Sep 17 00:00:00 2001
From: sub77 <sub77@ymail.com>
Date: Fri, 19 Jun 2015 04:22:27 +0200
Subject: [PATCH] Add support for matissewifi, matisselte, matisse3g, falcon

---
 kexec/arch/arm/mach-shamu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kexec/arch/arm/mach-shamu.c b/kexec/arch/arm/mach-shamu.c
index d1fb092..786567e 100644
--- a/kexec/arch/arm/mach-shamu.c
+++ b/kexec/arch/arm/mach-shamu.c
@@ -210,7 +210,7 @@ static int shamu_add_extra_regs(void *dtb_buf)
 }
 
 const struct arm_mach arm_mach_shamu = {
-    .boardnames = { "shamu", NULL },
+    .boardnames = { "shamu", "matissewifi", "matisselte", "matisse3g", "falcon", NULL },
     .choose_dtb = shamu_choose_dtb,
     .add_extra_regs = shamu_add_extra_regs,
 };
